diff --git a/g2o/solvers/CMakeLists.txt b/g2o/solvers/CMakeLists.txt
index a6c92b7..c86a5d2 100644
--- a/g2o/solvers/CMakeLists.txt
+++ b/g2o/solvers/CMakeLists.txt
@@ -12,6 +12,9 @@ if(CHOLMOD_FOUND)
 endif(CHOLMOD_FOUND)
 
 # Sparse Module of Eigen is stable starting from 3.1
+if (DEFINED Eigen3_VERSION)
+  set(EIGEN3_VERSION ${Eigen3_VERSION})
+endif()
 if (DEFINED EIGEN3_VERSION)
   if(${EIGEN3_VERSION} VERSION_GREATER "3.1.0" OR ${EIGEN3_VERSION} VERSION_EQUAL "3.1.0")
     add_subdirectory(eigen)
