diff --git a/CMakeLists.txt b/CMakeLists.txt
index 32a9c7a..d233953 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,15 +3,37 @@ project (valijson)
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
-option(valijson_INSTALL_HEADERS "Install valijson Headers." FALSE)
 option(valijson_BUILD_EXAMPLES "Build valijson Examples." FALSE)
 option(valijson_BUILD_TESTS "Build valijson Tests." TRUE)
 
 SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
 
-if(valijson_INSTALL_HEADERS)
-    install(DIRECTORY include/ DESTINATION include)
-endif()
+include(GNUInstallDirs)
+
+add_library(valijson INTERFACE)
+add_library(valijson::valijson ALIAS valijson)
+target_include_directories(valijson
+    INTERFACE
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+)
+include(CMakePackageConfigHelpers)
+
+configure_package_config_file(
+    valijsonConfig.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/valijsonConfig.cmake
+    INSTALL_DESTINATION ${CMAKE_INSTALL_DATADIR}/valijson
+    PATH_VARS CMAKE_INSTALL_INCLUDEDIR
+)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/valijsonConfig.cmake
+        DESTINATION ${CMAKE_INSTALL_DATADIR}/valijson)
+install(DIRECTORY include/valijson DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/valijson)
+install(DIRECTORY include/compat DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/valijson/compat)
+install(TARGETS valijson EXPORT valijson-export)
+install(EXPORT valijson-export
+    FILE valijsonTargets.cmake
+    NAMESPACE valijson::
+    DESTINATION ${CMAKE_INSTALL_DATADIR}/valijson
+)
 
 if(NOT valijson_BUILD_TESTS AND NOT valijson_BUILD_EXAMPLES)
     return()
diff --git a/valijsonConfig.cmake.in b/valijsonConfig.cmake.in
new file mode 100644
index 0000000..5ec282f
--- /dev/null
+++ b/valijsonConfig.cmake.in
@@ -0,0 +1,3 @@
+@PACKAGE_INIT@
+
+set_and_check(FOO_INCLUDE_DIR "@PACKAGE_CMAKE_INSTALL_INCLUDEDIR@")
\ No newline at end of file
